kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java

```diff
@@ -0,0 +1,193 @@
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal;
+
+import java.util.*;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.delta.kernel.*;
+import io.delta.kernel.engine.Engine;
+import io.delta.kernel.types.StructType;
+
+import io.delta.kernel.internal.actions.*;
+import io.delta.kernel.internal.fs.Path;
+import io.delta.kernel.internal.replay.LogReplay;
+import io.delta.kernel.internal.snapshot.LogSegment;
+import io.delta.kernel.internal.snapshot.SnapshotHint;
+import io.delta.kernel.internal.util.SchemaUtils;
+import io.delta.kernel.internal.util.Tuple2;
+import static io.delta.kernel.internal.TransactionImpl.DEFAULT_READ_VERSION;
+import static io.delta.kernel.internal.TransactionImpl.DEFAULT_WRITE_VERSION;
+import static io.delta.kernel.internal.util.SchemaUtils.casePreservingPartitionColNames;
+import static io.delta.kernel.internal.util.VectorUtils.stringArrayValue;
+import static io.delta.kernel.internal.util.VectorUtils.stringStringMapValue;
+
+public class TransactionBuilderImpl implements TransactionBuilder {
+    private static final Logger logger = LoggerFactory.getLogger(TransactionBuilderImpl.class);
+
+    private final long currentTimeMillis = System.currentTimeMillis();
+    private final TableImpl table;
+    private final String engineInfo;
+    private final Operation operation;
+    private Optional<StructType> schema = Optional.empty();
+    private Optional<List<String>> partitionColumns = Optional.empty();
+
+    public TransactionBuilderImpl(TableImpl table, String engineInfo, Operation operation) {
+        this.table = table;
+        this.engineInfo = engineInfo;
+        this.operation = operation;
+    }
+
+    @Override
+    public TransactionBuilder withSchema(Engine engine, StructType newSchema) {
+        this.schema = Optional.of(newSchema); // will be verified as part of the build() call
+        return this;
+    }
+
+    @Override
+    public TransactionBuilder withPartitionColumns(Engine engine, List<String> partitionColumns) {
+        if (!partitionColumns.isEmpty()) {
+            this.partitionColumns = Optional.of(partitionColumns);
+        }
+        return this;
+    }
+
+    @Override
+    public Transaction build(Engine engine) {
+        SnapshotImpl snapshot;
+        try {
+            snapshot = (SnapshotImpl) table.getLatestSnapshot(engine);
+        } catch (TableNotFoundException tblf) {
+            logger.info(
+                    "Table {} doesn't exist yet. Trying to create a new table.",
+                    table.getPath(engine));
+            schema.orElseThrow(() -> new IllegalArgumentException(
```

Comment:
Depending on how we expect the connector to implement creating new tables could this be a user-facing exception? i.e. I try to append data to a non-existent table. Is it the connectors responsibility to check if the table exists beforehand?